lightclient/src/background.rs

```diff
@@ -46,29 +50,160 @@ pub enum FromSubxt {
+                method,
+                params,
+                sender: tx,
+            })
+            .map_err(|_e| LightClientRpcError::BackgroundTaskDropped)?;
+
+        match rx.await {
+            Err(_e) => Err(LightClientRpcError::BackgroundTaskDropped),
+            Ok(response) => response,
+        }
+    }
+
+    /// Subscribe to some RPC method via the background task.
+    pub async fn subscribe(
+        &self,
+        method: String,
+        params: Option<Box<RawValue>>,
+        unsubscribe_method: String,
+    ) -> SubscriptionResponse {
+        let (tx, rx) = oneshot::channel();
+        self.to_backend
+            .send(Message::Subscription {
+                method,
+                params,
+                unsubscribe_method,
+                sender: tx,
+            })
+            .map_err(|_e| LightClientRpcError::BackgroundTaskDropped)?;
+
+        match rx.await {
+            Err(_e) => Err(LightClientRpcError::BackgroundTaskDropped),
+            Ok(response) => response,
+        }
+    }
+}
+
+/// A background task which runs with [`BackgroundTask::run()`] and manages messages
+/// coming to/from Smoldot.
 #[allow(clippy::type_complexity)]
 pub struct BackgroundTask<TPlatform: PlatformRef, TChain> {
-    /// Smoldot light client implementation that leverages the exposed platform.
-    client: smoldot_light::Client<TPlatform, TChain>,
-    /// Per-chain data.
-    chain_data: HashMap<smoldot_light::ChainId, ChainData>,
+    channels: BackgroundTaskChannels,
+    data: BackgroundTaskData<TPlatform, TChain>,
 }
 
-/// The data that we store for each chain.
-#[derive(Default)]
-struct ChainData {
-    /// Generates an unique monotonically increasing ID for each chain.
+impl<TPlatform: PlatformRef, TChain> BackgroundTask<TPlatform, TChain> {
+    /// Constructs a new [`BackgroundTask`].
+    pub(crate) fn new(
+        client: SharedClient<TPlatform, TChain>,
+        chain_id: smoldot_light::ChainId,
+        from_back: smoldot_light::JsonRpcResponses,
+    ) -> (BackgroundTask<TPlatform, TChain>, BackgroundTaskHandle) {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let bg_task = BackgroundTask {
+            channels: BackgroundTaskChannels {
+                from_front: UnboundedReceiverStream::new(rx),
+                from_back,
+            },
+            data: BackgroundTaskData {
+                client,
+                chain_id,
+                last_request_id: 0,
+                pending_subscriptions: HashMap::new(),
+                requests: HashMap::new(),
+                subscriptions: HashMap::new(),
+            },
+        };
+
+        let bg_handle = BackgroundTaskHandle { to_backend: tx };
+
+        (bg_task, bg_handle)
+    }
+
+    /// Run the background task, which:
+    /// - Forwards messages/subscription requests to Smoldot from the front end.
+    /// - Forwards responses back from Smoldot to the front end.
+    pub async fn run(self) {
```

Comment:
A simplified version of what was called `start_task` before. No longer takes in any args, so the API for `BackgroundTask` is super simple; just `.run().await` it to start it running (and in the future we can return something users can poll which will just wrap this).